appinventor/components/src/com/google/appinventor/components/runtime/SpeechRecognizer.java

```diff
@@ -103,11 +126,67 @@ public void BeforeGettingText() {
   }
 
   /**
-   * Simple event to raise after the VoiceReco activity has returned
+   * Simple event to raise after the VoiceReco activity has returned.
+   * Partial is 'false' when SpeechRecognizer stops automatically after listening,
+   * else it is 'true' if SpeechRecognizer returns partial results.
    */
   @SimpleEvent
-  public void AfterGettingText(String result) {
-    EventDispatcher.dispatchEvent(this, "AfterGettingText", result);
+  public void AfterGettingText(String result, boolean partial) {
+    EventDispatcher.dispatchEvent(this, "AfterGettingText", result, partial);
+  }
+
+  /**
+   * Method from SpeechListener interface.
+   */
+  @Override
+  public void onPartialResult(String text) {
+    result = text;
+    AfterGettingText(result, true);
+  }
+
+  /**
+   * Method from SpeechListener interface.
+   */
+  @Override
+  public void onResult(String text) {
+    result = text;
+    AfterGettingText(result, false);
+    onDelete();
+  }
+
+  /**
+   * Method from SpeechListener interface.
+   */
+  @Override
+  public void onError(String message) {
+    result = message;
+    AfterGettingText(result, false);
   }
 
+  @Override
+  public void onDelete() {
+    speechRecognizerController = null;
```

Comment:
I am currently unable to understand when this onDelete() method is called. Please explain a bit.
Also what kind of changes can happen in designer while speech recognition is running? Can you please provide an example for the same.

Reply:
After reviewing the code, I realize there is a mistake here. onDelete is unreachable in the current system. Instead, we should be implementing OnClearListener instead. The underlying problem is that when one changes the contents of the designer, we send the code to rebuild the entire screen and all of its components. If in the companion app you've started speech recognition, and then make a change to the designer (say, change the color of a button), the screen will be created and a new SpeechRecognizer initialized while the one from the previous version of the screen is still running. Eventually, it will get GCed, but in the mean time there is no way to interact with it so it will continue to record speech.

Reply:
Note when you switch to using OnClearListener, you also have to call form.registerOnClearListener(this) in the constructor to register the component.